src/expr/src/relation/func.rs

```diff
@@ -748,27 +820,113 @@ where
                     if offset > start_offset {
                         Datum::Null
                     } else {
-                        datums[end_idx].0
+                        args[end_idx]
                     }
                 } else {
-                    datums[end_idx].0
+                    args[end_idx]
                 }
             }
             // Forbidden during planning
             WindowFrameBound::UnboundedPreceding => unreachable!(),
         };
+        results.push(last_value);
+    }
+    results
+}
+
+/// Executes `FusedValueWindowFunc` on a reduction group.
+/// The expected input is in the format of `[((OriginalRow, (Args1, Args2, ...)), OrderByExprs...)]`
+/// where `Args1`, `Args2`, are the arguments of each of the fused functions. For functions that
+/// have only a single argument (first_value/last_value), these are simple values. For functions
+/// that have multiple arguments (lag/lead), these are also records.
+fn fused_value_window_func<'a, I>(
+    input_datums: I,
+    callers_temp_storage: &'a RowArena,
+    funcs: &Vec<AggregateFunc>,
+    order_by: &Vec<ColumnOrder>,
+) -> Datum<'a>
+where
+    I: IntoIterator<Item = Datum<'a>>,
+{
+    // Let's create a new RowArena, to avoid flooding the caller's arena with stuff proportional to
+    // the window partition size. We will use our own arena for internal evaluations, and will use
+    // the caller's at the very end to return the final result Datum.
+    let temp_storage = RowArena::new();
 
-        result.push((last_value, *original_row));
+    let has_last_value = funcs
+        .iter()
+        .any(|f| matches!(f, AggregateFunc::LastValue { .. }));
+
+    let input_datums_with_ranks = order_aggregate_datums_with_rank(input_datums, order_by);
+
+    let mut encoded_argsss = vec![Vec::new(); funcs.len()];
```

Comment:
This is a Haskell naming convention for keeping track of collection nesting depth: when naming a nested collection, we add an s at each level. Like,
- if I have a variable that holds a single argument, it's named `arg`,
- if I have a variable that is a collection of arguments, then it's `args`,
- if I have a variable that is a collection of collections of arguments, then it's `argss`,
- if I have a variable that is a collection of collections of collections of arguments, then it's `argsss`.

The last one is the case here, because `encoded_argsss` is a `Vec<Vec<Datum>>`, where each `Datum` can be a record of multiple arguments. (The outer Vec is over input rows, the inner Vec is over the different constituents of the fused function, and the Datum is the arguments to a single constituent function, e.g., lag's 3 arguments.)

Haskell example: https://stackoverflow.com/questions/66066788/i-am-struggling-to-understand-how-concat-is-read-in-haskell